//http://adventofcode.com/2017/day/9

fn parse(input:String) -> usize {
    let mut groupscore = 0;
    let mut stream: Vec<&str> = input.split("").collect();
        //two empy strings are the result of this vec at the beginning and end
        stream.remove(0);
        let last = stream.len()-1;
        stream.remove(last);
    
    let mut stack: Vec<&str> = Vec::new();
    let mut skip = false;
    let mut garbage = false;
    
    for &chr in stream.iter() {
    if !skip {
        if chr == "{" && !garbage {
            stack.push(chr);
        }
        
        if chr == "<" {
            garbage = true;
        }
        
        if chr == "!" {
            skip = true;
        }
        
        if chr == ">" && garbage{
            garbage = false;
        }
        
        if chr == "}" && !garbage {
            groupscore+=stack.len();
            stack.pop();
        }
        }
        else {
            skip = false;
        }
    }
    
    groupscore
}


fn main() {
    //test 1, 1 groupscore
    let input = "{}";
    println!("final score: {}",parse(input.to_string()));

    //test 2, 6 groupscore
    let input = "{{{}}}";
    println!("final score: {}",parse(input.to_string()));
    
    //test 3, 5 groupscore
    let input = "{{},{}}";
    println!("final score: {}",parse(input.to_string()));
    
    //test 4, 16 groupscore
    let input = "{{{},{},{{}}}}";
    println!("final score: {}",parse(input.to_string()));    
    
    //test 6, 1 groupscore
    let input = "{<a>,<a>,<a>,<a>}";
    println!("final score: {}",parse(input.to_string()));

    //test 7, 9 groupscore
    let input = "{{<ab>},{<ab>},{<ab>},{<ab>}}";
    println!("final score: {}",parse(input.to_string()));
    
    //test 8, 9 groupscore
    let input = "{{<!!>},{<!!>},{<!!>},{<!!>}}";
    println!("final score: {}",parse(input.to_string()));
    
    //test 8, 3 groupscore
    let input = "{{<a!>},{<a!>},{<a!>},{<ab>}}";
    println!("final score: {}",parse(input.to_string()));
    
    //challenge, had to replace "s in vim ... 
    let input = "{{{{{{{<!>!>!<u,!>!!!!!!!>!!!>,<<!!a>},<!!!>,<!>,<}>},{{<>},<>}},{{{{}},{{{},{{<!>,<!>,<,'e!!'i'!!<,u>}}},{<ia!!!!!>a!>},<!!u!!!!!>,<a!<!!!>>,{<!!{!!',,u!!!>oo'>}},{{<oa!!!o{a>}}}},{}},{{{<!!a!>,<!a!o!!!>!!!{!a!!!>},<o>,{<!i<{}!!!><>}},{}},{},{{{},{<,!>},<!>,<!>},<,!>,<!}!!!>>}},{<{!>,<{oi!a{a!!e!>},<o!!!>>},{<e!','a!!<!i{a},!!!>!>},<>,<{}}!!!>ao!!i}ao!!!>,<i!!>}}},{{{<!!!!<{!,e!!!>>},{<ue!>!!!>o'>}},{{{<o'aae!!!>e!>,<a!!o!!!>!>},<>,{<}au!}o!!u}!>},<!!!>oa<!!,>}},{{{{<'<'!!{!!aoeu!!!!ui>},{<,!<!!!>,<!!!>},<!}}!!{!!aoua!>,<>,{<<,,!!!>i!!<<!>!!!>,<>}}},{<<a<au!'!!!!!>!>!e!!uae{!!!>},<>}},{{{<!!!!i!!!>!>},<!!!>e,!>!!!!!>!!'a{}>,<a!>},<!!}'!>,<o!!!><>},{{<,!>},<}!!,i!>},<'a<!!!>,<,!!ei'!!a!>},<i>},{{{<{e!!!>,<u!>,<!>},<'!!!>a!!!>}a!!>},{<!!,>}},<!!!>},<,u!!!>,<o>}}},{<!>},<a<!>},<o!>},<!!!>!>>,{{{}},{<!>},<a!>!!}ea!>,<!>},<{a{!!},!!!!!{a<'<>}}}},{{{{<!>,<!!!>o!!}!!!>a!>,<i}}!o!>,<ae!>>}},{{<'i!>e!!!>},<!>!>,<a!!{u!>>,<o!!!>>},{<!>,<>}}},{<o'!>!o!!!>!!!!u!}',!!{!!!>}!!!!!>>}}},{}}}}},{},{{{{{<!!!>{<'!io<,>},{{{}},{{<!!'u<ae>},{{<!>},<!>},<}!>,<}!!{!!!!u,!<!>,<i!<a}<>,{<e!>ai!!'!!a!>},<!!!>o<i!>},<o!!i!>>}},{{}}}},{{{<!>a,a!e>,{<!!!>o!!iu!>},<{!>},<!>io!!{!>,<i<>}}},<u!>,<!!!!<a,}!>'!!'ai>}},{{{<u<!a!>},<!!aa,!!!>,<!>},<eu{}>}},<!!e}!!'!!!>!i!!{!>},<!!!!!>,<<>}},{{{<a!>!>ua!!!>!>},<!!u!>!a!!>,<!>},<{!!!!>},{<!!!>,<!!!!{!>,<!>},<!!a<}!>{,i{'!>,<!>,<>,<{<'<!!!>,<eo}>},{<!!<!!>}},{{},{{<a!>!>},<a,,!>},<}!>!e}<<!a,>}}}},{{{<o'i>,<!aaia!!!>,<!>,<!!!!'!u>},{{{<a<!!!!!!a!!!>u!>,<!!!>>}},<ia!!a>}},{{},{{<!,u!>},<}!!u!!{!>,<>},{{<!>},<}{o}!>},<!>},<ui!>,<!!<!>!>},<>,{<'!!'o}}i!!!!!>!!{<,e>}}}},{{{{<o!!!>!!!>},<>}}},{<oo!>i!>},<!!!>!>},<}i'<!i,!>},<!!'>},{<a,o!!!>a!!!>>,{<aae!u<!>},<e!>}o!!!>}!>a'ia>,<a!,i,e!'!!e!>,<!a!!!>i!>},<>}}}}},{{<!>uu<a>,{<'}!!!>!!!><{!!{a,!>''o>}},{<u!>},<{!!!>{!!!>,<o'o{a>,<o!{a{!{!>},<!>,<ei>},{<!!!>,<!{}a!!e<>,{<!a!!i<ua!!o!!!>,<!!a!>}!>},<i'}a!>!>>,{<!>,<<!>aa{aa!>e!>,<<<>}}}}},{{{{{<a,!!!!!>!>e}!!,i}a,!!u}e,a{>}}}},{<!>,<>,{<,aua!!!>a!!e!!!>>}},{{<ie!!ea!><!>,<'>}}},{{},{{{{{},<!!e>}},{{<{!!a{}e!>}u!!!>u!a!>},<ue>}},{}},{{<aau!>,<!>!>,<}a!!!!a!o!!!!!!}<!!{!>,<'!><a>,{<!!{!{!>,<e!!',!!!!eue!>>,{<a!>,!!i}!!'}!!}!!ueu!!{}!!!!!>{u>}}},{{{{<!!,oo!>,!!}a!!a!!!>!>>,{<!>}<a!>},<!!!><!!!>,<ue<!>'>}},{<{,!>!'a<<!>,<a!>!!!>!<}<<>}},{<{!!!!!>},<,!>'<!>e}!>,<i!!!>'!!>}},{<!o!!!!!>o!>},<u<!>,<!>{}!!a!!a>}}},{{{<eo!>,<!!a}ee!>},<ua!>},<!!!>iu>}},{{<}'aa!!!a!'{!!!>!!!>!>!>o!a!!}!!!!!!!>,>},{}}},{{{{{<!!!!!>!,<}!!!!!>,<a!<<a!!!>>,<!!u!>e!o!!!!!>!>},<!>e!!!!!>io!ie'!!>},{<!>,<a!>},<<ou!>},<ei!>,<i!!!>,<<{o!!!!!>{>},{<!>,<ao>}},{{{{{<<o!>,<aa!>o!,o!!!>!aai'!!'>}}},{{<!!aea'}aei!!a!!!!<iaa,>},{<>}},{{{<!>i!}e,io<u!!!!!>,<!>},<i!!}i>},{<!aa!!a}!!}>}},{{<>}},{{{<!u!>!},ea!!!>ue,!>!>{!>},<{'}i!>>},{<'a,a!>a>,<!!!>!>,<!!iaa!i!!!!!>,<{>},{<ie}}'aa!>,!>,<>,{<>}}}}}},{{<}!{i!!}!!!>i,!>,<}aa!!i!o>,{{<e}>},{<ie!'!!<!>},<i!!!>,<!!u!!{>}}}},{{{},{<o!>,<!>,<!>,<!!!>!!!!!!}!>{!!i!>oe'>}}},{{{<!!e!!o{!>,<ei!>,<,u<o!!!!!><>,<!!e!!!!!>>}},{{{{}}},{{<a!!!>},<o<!uo!>'!{!>},<,},!>,<e<>}},{{<!aia>},{<a!>a'{<>}}}}}},{{{}},{<!!>,{<!>},<!!!>},<{!}e>}}},{{<,!>},<!!ai'i!a!>,<!>},<a{au!>>,<<!>,<!>},<!!'!a!>},<!!!>!>},<'!!!a}'>},{{},{},{<,!!e!>},<!>},<!!!>!>},<!a!>},<a!!o!>u,<,,!!>}},{{<<a!>a}!a!!!>},!!e'{!!!!'!!!>!!!!o>},{{{<,>},<}!!!>a>},<io!!a<i<{}!'!>,<,{!>>}}},{{{<!>!>!>!!a>,<!>!!!>!!<'}'>},{{<o!!{a!!!>'a!!>,{<!!!>o!!!>,<o!!!!i{<'}!!!<e,>}},{}},{{{<a!>o<o!>u!}i!>},<!>,<,a!!a,''!>},<>}}}},{{{{{{{<<!>},<>},{<o!!!!!>},<!>''o}o!!<au!>},<u!!!>e!!!!!>>}},{<!>,<a,ae}',{}ua<u!>},<,eu>}}}}},{{{<!!!>,<}'<<{{!!e!!!>{i}!>},<!>,>,<}!!<i!!!>},<{!>},!!!!i!<!'!!,!!{u<!>>},{{<a!>,<'a<!!!!!>a{!>>,<ea}a!o<<'ui!!<o>},{{{{<}o>}}}},{{}}},{{},{}}},{{<a!a!>},<}!!!!!!!>u!>,<<a!!iao,o!,!!!>>},<!!!>!!!!ai!!!><!!}!>a!>a'ae>},{{{<e,!!'!!!,!>,<}au!>,<!><ai',<a>,<!>,<{a'}i!>},<}!>,<,>},<>},{{<'e>},{}},{{{}},{<e!!i>}}}},{{<,>,{<ou!!!>!,}'ia!!e>}},{<o!u!!!!u!!!!<!!}!>},<<!>,<!>,<!!{!!ue!>eie>,{}},{{{<!o!>'o!!u!>a}'!>!!!!!>!>},<}a>,<}a,e!!!!!>a!>},<!>>}},{<a!>,<i!!!>,<!>u!}!!u!!>},{<},!!,ioi!!!!!>!>'!!!>u>}}}},{{<!!!oa>},{{{<{a!>},<!!!>!!i{{ua>,{{}}},{<i<{,,!!}!!u!>}u>,<!!iu}!!<ua!!!!!>>}},{{},<!!>},{<!>},<!!!>!>},<i}<!!!>!>},<{,>}}}}},{{},{<!>!>!{o!!!>!>},<!>},<<!>,<{!!'!!!><!!u>,<a'>}},{{{<{!>!>!>!>!>!>},<!>e!!!!u!!!!!>u'>},{},{{{<>}},{<!!!>{<!>,<{e!!{!!!>au!>aa!!!>,<>,{<!>a>}},{{},<oa!>>}}},{},{{{},{<{e'!>,<ioo!!!!!>,!!!>>}},{{{<,{a!!!>},<ea'e,iu>},<u!!!>!!!,!!iuuueaaua{>},<}>}}},{{{{<'!>},<e!>!a!!!>!!!!a!!{a!{!ieii!!!>!o>},{{<!>,<!>o{!!!>},<!!!><!>},<i!!!!>}}},{{<a!!'o!!>},{{<!!!!a!!ia!!<}eoua!'!>},<'>}}}},{{{<!!!!,}!!!>},<!!!!!>,<>}},{{<{!!!!!>a!>},<!>,<i!>},<!>,<!>a!>},<!>,<{,!!!>!!!!uo>},<,!!u!}>},{<o!!!!uo'e!}a<o!!!,<!!oa{>,{}}}}}}}}},{{{{},{{<<<>},<!!!>{!!!!i{!!!>oo>}},{{},{{{<ea'ue!>},<!>{!}>,<,>}}}}},{{<>,{}},{{{<!!!>'}{o!>,<!!!>!i!!!>,<!>,<!!'a!!!>!!!>}>},{{}}},<!!!>a!!!!<!!'!u<!>,<>}},{{{},{<!!,!!{!!!>},<!!!>>}},{<!!!!!!!>o!!>,{{<u}i!>,i!!,i}e>},{<!>},<!!a!}!>,<!a!>},<!!!>!!<a!!a!>},<!!!>>}}}},{}},{{{{{<e{!!!><}ei},ea!!!>!>},<'>,{<<!!!!!>,e!e}!!!>},<i'ua!>,<>}},{{}},{<a!i!>!!{!!{o!>},<!i{!>,<>,{<ei,>}}},{{{<!!!><!{a'oo>}},{<a!>a!'e!>,<!>,<>}},{{{},<!>},<!!!i!!!>},<!>,<,o!>},<}!!!>{!!e!!!!u!!<<!!}!>},<!>,<}>},{<ae!!!>{>}}},{{},{<!>!>,<!'!,ae!>o,!!!>!>,<e,}!!<o,>}},{<a{!>,<<i!!!!o{<o>,{<!!}>}}},{{{{<!!!>}!>,<ia!!iaeoieaaii!!!>u>},{<!!}e!!e>,<!!!!<aie!!o!>},<!>,<!>,<!!!!!>!!<!eu,!>u>}},{},{{<!'!>,<<!auo>},{<!>,<>}}},{{<!>},<'a!a!!!!{!>!>,<o!>},<u!>!{}!>>},<o'e{'aa!!!>>}},{{<},a!o!>},<,,!!!>,<!>!!!>!!!>a!>>,{}},{{<<eau>},{{{<!!,>,<>},{{<!!!><!}a!i!!}!>},<o!!a''e>},{}},{{}}},{<a!>},<!>,<!'!a{!!!!!>},<!>},<>,{<!!!>e!!!!!>au!>},<{ue>}},{{{{<o,!>!!!>!!ua!>!!!>o!}}>},{}},<i!>},<<e!u{eua!e'},>}}}}},{{<ee!!!>'i!>auu!!{{!>a>,{}},{{<u'!<!>},<!>},<!!!>'!!>},<a!!!>ae!}!>}>}}}},{{{{<!!!>!a<,!!,ua{!}'a'!a!>,<!>,<>},{{<a'!{,aa{u!!{<!>u!>{i!>,<!!!>,a>},{{},<!>!!!>!!!>},<!'!!,i,ii!!!!!!!><'!!!!oi,!!>}},{{{}}}},{{<,!!!>,<,u!!!'ia,>,<!!i!>},<!!{<>},{{{{<!!!>i{e!>,<i,!>},<u!>,<!a,o!>},<>}},{<!>,<e'>,<ee!!!>a!!!>},<<!>>}}},{<o{<e!!!!!>!!e'!>,<au>,{<!!!>{a!!!!!>!!!>au{a>}}},{{{<!!aeo!!a}<{'>,{<u>}},{{<e,,>,{{{<o!!!>o!a!>{{{!i'!><!>,<!>,<>},<a'!!!!!,u!!e<!a,>},<a!!ue>}},{},{}}},{},{{{<!>a!!!>!!!>!!oo'o!!<}{ia!>,<'u!>},<e>},<ia!!!>ou'!>,<!!!>},<a<<{i>},{{{{<!!!>o<<a{!!}a'a,,!!!>a!!!><>}},{{{<!>,<eau>}},<eee},!!eo!<}>}},<u!!}oe>}}},{{}}},{{{{{<ai!>,<'{>}},{<e!!!>},<!>,<!><!!!>a>}}},{{<!!<!!!>},<!>},<a<<!>!a!>,<e{>,{<<a!>,<>}},{{<,o!>a!<u!>,<!>,<u}!!!!>},{{<o'>},{<,!!!>!o>}}},{{},{{<,!>!>},<>,{<!!!!!!ao!!a!!a!>,<!>},<>}},{{{<!!}!!!!!>o!!!>{!!>}},{{<!!!!!>,<e'u!>},<'u!>},<u!>!!e,ea>,<o!a>},<euoo!!!>>},{{{{<!!!>'o{>},{{},<e!><>},{{},{<<e,!!!>!>},<}!>,<!>},<'{ee!>,<e,>}}},{<{{<a{u<!!!!i!!}!>,<''{a!!!>>},{{<!!i!>,<a!!{'<e!e{!!!>,<<{!>},<!>!{!!!>!>},<>},{<,iu!!'},ao!!au>}}},{<!!!>i,>},{<ee!>i<!!u{!!!>oa'}>,{<,e'>}}}},{{<!>},<<,,>,{<}>}},{<u}!>,o{!<<,!>,<e,!!a>}}},{{<!a!>,<e!>,<o!>!>},<a}!!eu!>},<!!<>,{}}}}},{{{<>}},{},{<a!ui!>},<!{<!<!uia'!>u,aa!!!>'>,{<!!i!>,<aeae}a>}}},{{{{}},{{{<!>},<aei!o!!!>,<{!>,<!!!>e,a>},<!!<>},{<!>aaii'e{e,!!!>,!>,<{>,<!>!!!>,<o!!}{e!!!>i!>,<!>!!!>!<a<{!>},<!>},<>},{{<!>},<!>'ea!!!!!>,a}',a!!!>!!e!!uu>}}},{}},{{{{}},{}},{{}},{<!>},<}!>},<!>},<!>,<o!<!>au{{ea!>,<!}!>>,{{<<<!>},<o,!!!>{>}}}}}},{{{<<!>},<!>,<>},{{{{<!>},<a!!u!>}!>,<,!>,<!>,<,,>},{<!!!>{!>},<a!>,<!!!!!i,!>},<!!!>},<},i!><>}},{}},{{<!>!!!>a}!!a{!>},<!!ua!!{e}au!!>},<i<{!!!>e'!!!>eu{aaaa{a!!!>!!!>>},{{{}},{}}}},{{{<!!'!>o!>},<i!ai',{a>},<>},{},{{<'i!!{i}<ii'{oi'a}u!!ei!!!u>},{<>}}},{{<'!!!>!!a<u!>,<}!i},a>,{{{<!>!>,<!!!>a!>!>},<!e!!!>!><a>},<!>,<e!!'u<<<!e!'>}}}}},{{{{{{<{',a!>},<>,<}!uaa!>,<<eia<ie>}}},{{{<<a,i'>},{<u'!>e<!!aa!!!>,<!,!'>,{<!!!>,<uu,!}!!,uia'e!!ie!!i!a!!!!{!>},<>}},{{<!u!!!>},<!>,<!>},<!>},<a>}}},{{{}}},{{<,<{e!!!o!>},<!!!>'i!>,<!!!>a}>},{<a!!!>},<i'!,!!{!>},<!!}e>,{}}}}},{{{<{!>!!!>,>}},{{<!>eo<'u!a,'iu<!i!}!>},<>},{<}<!>},<!>},<>}}},{{{<!!!!!>!!!!<'aa!<!>,<a!>},<o!!o>,<}e<ou!>oa!}!!!>!!}iu}!>>},{}},{{<>}}},{{<e'i>},{{<<!>!!,>}},{{<a>}}}},{{{{{{{<>},<a!!{!!!!!!!>u!!!>a!!,a'!}}!!'!!!>>},{<!>,<!>},<ua>}},{{},<!>!>},<!!a!'ee!>,<}eoiao!!!!!!!eu>}},{{<a!!,!!!>},<!!}!!!>i!!!>,<>}}},{<i{<!!a!!!>!>},<,!!!!'>,<!><!!,!uo!!}<{uoo,'!!!!!>{>}},{{<e>,{}},{{<a<!>},<!>u!>!>,<i!!a!!oeo}}oa!!a>}},{<!!!,!!a!>},<i!>!!!>},<!u<!!!>>,<{a!!au'au<ua!>,<,,>}},{{{{}},{<<ei{!>,<>}},{{{<,!a!!a!><}ai!!!!!>!>!!!>!,ui>},{}},{},{}}},{{}}},{{{{<!!}o}!!a}<!!}'!!aao}!!!!e>,<>},{{{<!!!>i!>,<!!!>!>u'!>,<,>,<!!a!>,<a!!,!!e!!!>aae!!!>!>>},{{<!e},eouua!!oooeoaoo!!!>>,{<eu,!>!>,<!!!>!>}{<!>,<!!!>!>!!a!!!,ai'a>}},{<o'!>},<!!!>e!!!!}''oo!>,<a<>,<}o!>'!!a!>},<!>},<a!}e!!!>!!},>}},{<}i,<!!<u>,<{!!!!<a}>}},{<a,,i,e{!!!>,<'i<!!!>},<>,<<!!!>}u!!ao!>!a>}}},{{{},{{<a!>},<<}'o!>,<ua,!!!><!>,<!'<ae!>>},<a!>,<uoo!'!!i!a!!!!!>!><'au!!!!!>,<a}>},{<!!!>,!>},<,!!o<i!>,<iu!!,i!!!,{>}},{{},{{<a!>},<'!>},<!ea!>!!,a>},{<!>},<>}}},{{{<a!!!>!!<!>},<a!!ou!>},<e,!!i!uo!!!>!!{'>},<!>},<o{}a}!>,<>},{{{{},{{{},{<!>},<!>},<}!>!!o<>}},{{{<!a'<a!!!!!>a!>,<e>}},<!!!>}!!a!!u!!!>},<}!!!>},<,a}<!!i!!!>>}}},<!!!>!>euo!!,u!!!><!>!!<eo!>,<a!>,<>},{{},{{{<!e!!<!>},<!!!>,!ou{!,>}}}}},{<!!!>!e!>ee!!a!!!!!>aia'{!>,<>,<<e!!!>{!>!>},<<}>}}}},{{{{{{<!!'!<>}},{<{!a<a!>}a{<e!!a!!},}'>}}},{{{{<!>!!'!>,<,u!!!!i!!a!>!!e>}},{<,{!>,<!!!>!a!>,<{!!!>!!>}},{{{<,ao!!<{!>{'!>},<u!!!>!!!>o!!!,!>,<a<>}}}},{{{{{}},<!!!!a!>},<!>},<!>,<o!!<!>},<!!o{!>,<<oau!!!!}>}},{<!!!>!>!>,<<!>e,>,{}},{<,!!!>!>,<}!!eao<{,!!aa!>,<!!!>!>},<oaa!!a>}}}}}},{{{{<o!!!>!>},<'!!>},{<}{'!>'!!a{!>u!'ao{aa!!>}},{{}},{{<!!!}!>},<{{!>!!u!i>},{{{}},{}}}},{}}}},{{{{{{<>},{<>}},{{{<!!!>,!!!>!!u>},{<!ua!!!>!!>,{}}}},{{<e!>},<!>},<!!{ia!!!!!>>}}},{{}},{<}!a'}!>,<<!o!!eu!>},<!!uu'!!!>a!>},<!!e>}},{},{{},{{<!o<!a'a!!u<!!iu!>!>},<<ai>},<>}},{{{<!{ii!!}i!>,<a>,<!>!>,<!!,>}},{<!>'o,!!i!!<oo!>,<u'!>i!!!>'>},{{{},<>},<i!!!!,}!!!!aue>}}},{{{{{<!!a!!!>aa!!!><aeuoe!a>},{<au,o}!>,<>}},{{{<!>!>,<a,a!!!!!>'!e!ua!>,<!!>}},{<!!!a}!>,<o!<{!{!>,<!!!>}a,a!!u!>},<i!!}o>}}},{<a>},{{{<a{e<>}}}},{{{{{{<o!>},<!>!!a!>},<{ao!,i<!!!'{>}}}}}},{{{<!!!!!>!>!{e}{!>},<!!!!!>,<,}oao!><>}}},{{{},{<u!!!!'u!>!!e!!ou'u'u}>},{{{<,aa{o!!!eua>},{<a!>'!!!>,<{!!!e!!!!ae}!>},<<o{{!!!>},<u{u!>>}},{<>,<}'i,!!!>,<!>},<>},{{},{{<a!>},<}!>,<o<!>,,!>,<ueu'!>,<>}}}}},{<!>,<!'uia<,u>,<!>,<<{a!>},<!a!>,<>}}}},{{{{{}},{{<e{{!ee!>,<a!>,<{!>!!,!!i'u!!!>,<!!!>!!!>>},{{<!!a>},<i<ee!'e{'!!'uoi'!>}>}},{{{{<a!!!!!>>}}},{{{<!>},<!!!>!>,<!!!>!!<!!!>!a!>,<<a!>,<!>!!!>!!!>{a,>},{}}}}}},{{{{<!!!!e!>,<<!!!>!>},<!}o>},<!!,'a,!!!>,!>,<!>},<!!o,!!!>},<!i<!>,<!>},<{>},{{<e!>},<!!a{!>},<>},<a!!>},{{<!>,<e',ae!!!>a!!<{,o!ai!!a}{!>>,{{<a!!!>,!>},<<!{!>},<u!>,<!>,<>},<!iu'!>},<!!i!!i!!!>!!!>!>!>,<i!!u,!><a>}},{{<o!,ee<>}}}},{{{{<}!>,<<a!>oaa''!>!!!!!>!!o>},{<'<,!<ao,!>},<!>},<ai!>!!<'o,!>>}}},{{<uea!>,<{'!>},<o!a,e!!u}{o!>},<{}!o>},{{<!>}<i!!o!>,<!uoo>},{<,!>!!!>'a!!!>!><,ouo>}},{{<ea,auoao!>,<a<>}}},{{{<!!ai,o''!!<e<{a<eo!!!>>,{<a!!!e}!!!>!!!u!!!>!!<!!!!!>},<{u}a>}},{<',!>,<o!>,<!>,<iu,!>,<>}}}},{{{{{<!>,<<!!!>!!{!!!!,>},<!!!ea!!!>a>}}},{{{{{<!>},<!>ue'u'!!a,!!!>}>}},{<!!ua!>,<!><!!,o!>,<}}{>}}},{{},{<ui!>},<!!}u!!!>,'!!uu{!!{<e<!!e!>,<!>>}},{{<i!!!>ua!!!>!!!>!!!>u!!'u!!!>},<!e>},{<!,!}!!e<!!'!>!!a!!!!!>!>},<,>}},{{{{<!!!>!!{}u!!!!ii!!!>aa'>,{<ie,!}!>i'!ee'!>,<a!e!!!>{>}},{{<a!!!>o!!!{o!>,<ae!>,<a!'{iaoai>,<!!,e>},{},{}},{{<u<'<}>},{{},<!>o'!!oao!>!!!>},<!{!>},<>}}},{{<!>},<!>i,!!o>},{{}}}},{{},{{<!!o!!iu'>},<!!!>{a<eao>}},{}}}}},{{{{<}{>},<a!!!>!>,<u'!>,<!>,<{iio,'!!uia>}},{{{{<!>,<i!!o>},{<{{a,'!!,!>,<!!}u!!a}a!!!>a<!>},<>}},{{{{<a!!!!a!e'!>,<!>'<!a!e<>}}},{<!,!>},<{ai'i>}},{{{<<!>,<a!a{ao>},{<,>}},{},{{{<iui!!!>,<<},!}io!!,o{>},{{<,a!!!!,e!>,<!>},<,!!!>e!>i!>,<u!!{i!!,!!!>},<>},<!}<}!!e!>,<!'{i!!!>},<!i!!!>!'{,!!i>}},{<u!!}!a!!!>!>,<!!!>}!!au!>},<a>,{}}}}}},{{{<>},{{<!iu<i!>{u,!o!>},<a!>},<>}}},{<<!>!>,<!!!!!>!a>,<a!!!>e!>,!>},<a!u!!!>>}},{}}},{{{{{{<u!!,aa<u!!'i,{>,<!!{>},{}},{}},{{{<uio!>,<'!{io{!>>}},{<!>},<!!o!>,<!',,}'!}i<e!o<oa!>,<!!!>e>,{{<<a!>i!!!>e'!!!>,<>}}},{{<'!i>},{{},<{o!,>},{{{<!>,<ei!>!!,<!>,<!>!>,<{a>},<!<>},{{<!>},<aa!!}o,!!!>u!!!>,o>,<uu!>!>},<>},<!eai!!!>!!!>,<i!>,<!}oa!}!!!>o',!>},<!>},<>}}}},{}},{{{{{{<}!!!'a!!!!oa'}!!ee!!,!>,<!aa>}},{{<'<!uua!!!u'!!uua,!ai!!,!>,<!!!>,<{>}},{<!u!!!>!o!!u>}},{{{<u,!!!!!>}!>},<!!!>!!a}!!!>o,'!>,<!>!}o!i>},<!>!>},<eeei'!>!!u!>,<!!a,,!o!>'o!!,'>},{}}},{{<!>,<{e}!!!>e>},{{<!!<e!!!>!!a>,<a'a!!<!>},<a!!i!>ea}!!<<!!i!>},<!!!>},<>},{<!!!>'!>},<auu!!!!!!!>!>},<'>}}}},{{{<!>,<eaa!>,<ia>}},{<!!!>a!{!,<!>},<'<!>o'!>a}!eea!>,<!!!!a>,{<!!!>!!<o{!>}!>},<}}!>,<,!e{!<e!>,<!>!aa>}}},{{{{{<o,!>},<i!>>},<!>,<,'>},{{<!><!o!a!!!>},<!!!!!>,a!,a,}!{a!>,<!>,<!>>}},{{{},{{{<a'a'!!i<!ea!!!>,<i,!!!>e{u!}>},<!!!><u{!>!!!!!>i{a!!aau>},{}},{{<e!>},<a!!!>!>!!!>>}}},{{},<!a,!>,<,ua!ao!eoa!o<>}}}},{{{{<!>},<>},{{{{{{<e!>},<ao},!>},<!{a>}},<i!!'!o!a!!!>},<a!>,<!!!!>},<!!i},<i!!!!<'!>,<'>}},{}}},{<!i,!i!>'}!!!!!>a{}'a!!!!>,{<ouia!!!!!>,<'!!!>',{!>!!}!!u!!<'!>,<!>},<>}},{{{{}},{<!>,<a}!!!>}>}}}},{{},{{<e}a!!}aaaa!!aa!!{!u!!!!!>>},{{},{{{{<!><!>!!!>u,!!i{!>},<'!>ao}{>,<!!!!a,o!!o'>},{{<'!!!>,!!eiu>},<}i!!!>,!!o}!!!!!!!>,<ao}u>}},{{{{{{<!>o<'!!e!!!{}!>},<!a>,<u!'!!u!>},<!>a}!e!!!>a}{!>,<>},<!!!>!!!>>}},{}},{{<!!!>e!>},<>}},{{{<!,!>!!o}<'}}>}},{{<!!!!!!!!'!!!!,'!!!>!>,<},a!!>},{<>}},{{<{ua>}}}},{{},<!a!!!>a}!!!>},<!>},<>},{{{<<>}}}},{{{<'a!!!>eouu!>,<!!!>,<uo!!a!!i>}},{}}},{<o!!!>a!>oe<!>},<<!!>,<!!!a,aui!>aae!>'i}!>},<!a>}},{<i!>},<o!!!>,<a!>},<>,<}au!>},<!!!>u!>,<u!!!>},<e!!}a!!''!{>}}}}},{},{{{{<<!!!!!>{{!!!>!!oe!>,<a!<o!>,<a>},{{<!ae!>,<>}}}},{{<!>!>,<{!!!>u!>'!>},<>,{{<o!>},<,e>}}},{<<eaoee'!!!>,<a!!!>!!e!!!>,<!!!!!>a!>o!>},<e>,{}},{<!>,<'!!!<!>'>,{{<!!a!>},<!!!>!>ua>}}}},{{<aa!>,<'>,<!!!>,<uo!>,<ou,!!!>!>},<!!!>,<'u!!!>!>,<!!!>,<u>},{{{<ua!>,<}u!>,<!!!>!!!>!!!>!!!!'>},{<u'e}!!!>!!!,!!o!u!>},<a<!!a!>,<>}},{{<!>},<<o{u'>},{}}},{<!!a!>u<ai!!!>!>,<!>}!!'o!>,<>,{}}}}}},{},{{{{{<iia!!eia!!!>aa>},{<o,!>e!>!!{!!!>a}i!>,>,{<<{!!a'!>>}}}},{{<<!!},!!>},{{{},{{{},{<!,a!!!>},<e!<!!!!!>u!>u>}},{}}}}},{{},{<e!!!>{,<a!>},<!{,e!o!!!>!!>}}}}},{{{{<'<!!!>},<!!!>!!!!{!!!>u>}},{<a>,{}}},{},{{{<o!>},<}!>,<!!!>{!!!>,<!!a>},{<u!!'i!<!>!!!!!>,<!>!!!>!>,<!!>}},{{<e!>,<u!>e>},{<!!{!!!aau<!!o!!>,{<o'!!!>!!!>!!!>},<ui!>,<>}}}},{{{<!!!}!!!>{o'!>!>aoa,o'<o!>a>}}}},{{{{<!a!!!>{i,oe'!!,au!><!!!>},<u,>},{{{<ei!!!>!'i'a<>},<!!,!}!!!><a!!}'au!>a!>,<!!!ae''>}}},{{{<{o}!>,<aa!>,<<,'{u,<}<!}>},<{u!!!>!!!>!!e<>},{<>}},{{},{{},{<u{!!!!>,{<!}u!>},<'e!!!>!>,<!>!!!!,!!!!!>!!!>!>},<!>,<!!}!,>}}},{{},<e!!<aeui!!eo>}},{{{<!'{!!!>'i{a!a'!!!!!!!>>},<{u!>,<}<ia!!!>i>},{{<!!}!>,<{!!a!>i!!,!>,<<!!aa!!>,<>},{<!!{e!!<}>}},{{<!!!>'!,>}}}},{{<o!!!>!>,<!!!>a!!!>!>,<'!>!>},<}!e'o!>,<a>,<!!!>,<!,a!,!>i}!{!>,<!>},<<<a<>},{{<!>,<<!>'<!!!><!!ao>,{<<,a}{a!!!>'!>},<<<!!!>o!!a!>,<i>}}},{{{},<aiaea}!>u{!!!>!!a{uoaa!>},<!!!>!>,<>},{{<{e!<}!<!!!>!!!>,!!'!!o{,!!!>,<{!a!!a>},{<!>,<}!>},<!u!>},<o!!<{u>}},{<o,!>,<i!}!!!!!>,<!!aa!>},<u<!!!>},<>,<{o!a!!!!e!>!,!>!i!>!i>}}},{{{},<!!aou},!!!!!!i!>u!>,<}u>},{<!!,!<!a!!ea!!!!!>o}!i}i{!!!>},<!>,<'!<'>,{<!u{u!>!>,<,!>},<!>{<a>,{{}}}},{{},{{<i!!!>>}}}}}},{{{{{<!>},<u,!!!ui!!!><!uu!>},<!!!>e!!!{o!!a>}},{{}},{<o'<!!<>}},{{},{{{{<!a!iia,!!!>!!ea!!u',!!!>>},{}},<!!!>},<ua!!u!!!uio!!!>a>}}}},{{{{{{<!!!!'!!!!io,e,!!i!!!>'u>},<!!e!>},i!>},<e!>!!!>,<',}ee!>>},{<a!a}!!oaoe!!!>iua{>,{}},{{{{{}}},{{<<!!{<!>euao>}}},{{{<a}>,{<!>ea!!!>{e>}},{<a}a!>},<'!>,<'!!!>i!o!>i}!!!>!a>}},{{<!>}!>},<!!!>{!>>}},{{}}},{{<!>,<a!,!>},<!>ua<!!!>},<au!'a!!,u{!>},<!!!>>},{<!>e>}}}}},{{{{}},{<!!!>!a!>},<!,!>,<!>!!!>!!!>u!!!>,a!,u!!!>!e}!!!>e>}},{{{{<!{!!}!>},<,'i<!>,<!{ee!!a>}},<!aa!!{!!!>,<}i'{!>},<',>},{{<}!>{!>},<a!,<!>},<u>}}},{<o}u!!<'<!>,<i}!!i!<!i>}}},{{{{<>}},{}}},{{{<e!>!!<ee!!!>}aa!!i!}!!a!!uai>}},{{},{{{}},{<'oa<e!!!!{}'!!!>,<ao<a!>>}},{}},{{<!!!>>,{{}}},{{<,!!!>uuu!>},<}{,u<}'{>}},{}},{{<i!!!!}u<!!!>'!>'!!!>,!!!>!!!>,<}o!>,<,i>,{<}!!!>'!>,<!!!>!!{!!u}'u}auo>,{<!>u}a!>,<euo,!>,<ea,>,{{},<!!uia!!!!,<a!>,<u<!!!!!>e!!>}}}},{{<,ie>}},{<e!!!>a!>,<!>ue!!!>,!!!!!>},<i!}!>!!!!!>i!>>,{{{<<aau,'!>ua!!!!{i!!{a>,{}}}}}}},{{{{<!>!!!!>},{{<!!a<{<!>},<>},<,i>},{<!!!>'!u'a!!{!!e}>,{{}}}},{{<!>,<!!e!!!}!>,<u}i'o,!!u>,<!!i!>!>,<!!!!!!!!!>>},{},{{<!!!!,!>,<ue!!!,u!e!!{!!!>!>,<}'e>}}},{{{<<>}},{{<o{oi!>},<a!!!!a!e!!!>!>,<{,}u'!!!>>},{<ia!!!>!>},<}a'!>}>}},{{},{{<'!>},<!!i!!!>!!!>},<!!!>u},'!>,<o<i{ou!!>,{}},{{<!>,<!!a{{>,{<{!,o{<!>},<!!ea!>,<,!>},<!>!>,<'a!>,<oi!}>}}}}}}},{},{{{<!>,<!>,<<!!!<!!!!!>,<u,ae>},{<}i!!au,'!>},<!>},<!!!!!!!>!>>,<a!>},<!<a!>,<!>,uio!>,<!>},<!!!>!!!><>},{}},{{{{{{<!>},<}a>},{<a!>,<!!!>>}},<{a!!!>!!!>!!!>},<}!!!>},<!!!>a'!!!!!>!>,<}>},{<!u!!!!!>!,<!!!>!>!!,<e{>}},{<!{!!}{o,u}a{!aa!>,!!!!}!!!>},<a>,<!>u!!!><a!>!!'},{oua!>},{!!!!!>>}},{{{<uua}<!!!>{,!!oe<>},<!>ue,!!!>!!!>},<!!u!!{e!!!>>},{{<aaa!!!!',,',!>},<u{a>},{<ei!!}!!!!!!!>{!!!>a!auo!>,<>}},{<!!e!>!!!>!>},<!>},<!>,<uoo!!!>,<iiui!!{!>,<!!!!!>,<!>},<>,{}}},{<a!!!>!!!>!>,<!>},<!!!>!>}a>,<'!{!!!>!!!>a!!!!!!!!o>}},{{{}}}},{{},{{{<io}a{{ea'!{<!!!>ui!>e>}}}}},{{{{{<{}<u!!!!!!!>o!!!>'!!}!!!uo!'!!!!!>>},<>},{<!!!>>,{}},{{{{{<>}},{<!a!!!>a!!oo!>}a,!!,uu!aa!!!>!>},<>}},{{<!!!>o>}},{{{{<>,<!>},<}!u!!!>,<{o}}a!>},<e!!!!ai!!!>,<}a>}},{{<i}!!!>{!!,iee!!!!!!{!>,<!>>}},{{},{<}!!!>!!!!!>{!>,<>,<!>},<e<!!,!!!!e!>,<!!!>>}}},{{},<!!!>},<<{ue}!oaaa!>oi<u!!i!>,<!i>},{{},{<!!u!'}i!!{!!!>,<!>,<!!!>},<!o'!!!>>,{<}!!!>!!!>,<aau!,>,{{},{<!o{!!!><{o>}}}},{{<>,{<!io!!'}e{!!!!!>!!!>!!!>!'}!!!>!aui>}}}}}}}},{{{{{},{<!ai!><ua!>!>},<!!e>}}},{{{<!>,<!>{{!>,<{}aia}!>},<!>},<<<{au>}},{<!<}!!a>},{{{{<!>,<!!!>,ai',!!!>iai!>}a>}}}}}},{{<!!!>}!!!<u'o','!>!!!>!>},<!!!!!>},<!>},<!!{>},{{{<'!i<}!!e>},{<!!!!!>oo!!<!u'!oaie!!!>!!!>ee{>}},{<>}}},{{{<o!!aa!!!!!!}{<e>},{<!!{!!!>>}},{<!!!>!>i!u!><!!!>e!!a!!!>i!!!>eu'>,{<i!!!>,<!!}!!!iu<a>}},{{},{<''u!>u!>},<,!!ia!!!>{'}'!>},<!!!>i>}}},{{<!>,<!!{!!!>!>,<!<}!>,<>,<!>,<!!!>!!{a!!ao!i!<<u!}!!!!!>,<>},{},{<!!!!!!!!ui!i!!oaaai}!'ue!!i!>,<ai!!!!!!!<>}}}},{{<!>},<}'!i'a}>,<!!!!!>,<{!ueo{},!!!>,<!!!>'!>},<o!>,<o,}>},{<'!!<>,{<!>},<<!!!>!!!>ee!!<{!>,<o{{{>}},{{{<!>!!,>}},{<!!e}!>},<{u>},{{{{<!>},<,!>i{a!o!>!>},<!>},<!><>},{{<!>},<eu!!a!>},<<!>!!ia>}}}}}}},{{{<!o!'aa}{!!{<aa>}}}}},{{{{<',eua!>{>,<a!!!!!!<a!>,<!!'}u!>,<>}},{<!>,<''!!<{!>,<}}o!>},<ae>}},{{<oi!!!>!!!>!>},<!!!!!>o>,{}},{{{{}}},{{{<!!o>,<!o>},{}},{{{<}u!>,<>},{<!!}a<!!!>uo{,!>,<!a!!{{}!!!>,<!!!!!>!>,<!>>}},{<a{!a{a!!!!,o!>},<iu{{!>a!{'o>,{{<!!!!'!!!>,<!a!!{,o!!!!!><'u>},{<!i}!!oa!!}<!!ee!!!>!!o!>},<!!a!>,<,>}}},{{<e!!}i}!i>},<o!>},<a'u!!!>a!!ea!!!>!>,<!>,<!iu{!'!>>}},{{<i!>,<!>,<ea>},<e<uo>}},{{{<!!!>e'!>},<''<eaue>},{{<}!!!o>}}},{{}}}}},{{<}!!!>},<u!!'e!!,!!!>,<!!!>!u!>,<!!a!>},<}a!>,<!>>},{{<!!!>!!o!>},<!!o!>,<}!!!>i>},{{<!>,<!>,<>}}}},{{{{<o{a}>},{<u!>,<,!!u>,<!>u!!!o<!io>}},{{}},{{{<a{a}'!>,<!}!!<!!!>,!!aa!!!>}i>}},{{<!>{!!!>!!!>},<!>,<!>!!!>',e>},<aa}a{e!>}{<ae<!>},<ae>}}},{{<!>!>},<!<a!!o>},{<!!e{!!aa<{i'!o<<!>},<!>,<!!!>o!!i!!!e>}},{{},{{{},{<a!>,<!!!>a>}},{}}}}},{{{{{<'!!!>ie!!!,a!>}'!>,<!!<e!a!>!>>,{<,,}!>},<ua!!'!!!>i>}},{{},{<!>,<<!!!>},<,!!e>}},{}},{{<,ou!o!!ae!!!>i{!>,<<>,{}},{<!!!!!>ao!>},<{!!!>aia!!!>>,{<{!>},<a!>},<!>a!!!>>}}}},{{{<!!}'ao<{u!i<!>,,!!<a!!!!ei>},{<!!!>a}i!>,<o!>},<!!!!!>!>},<>,<a!>!!!aua!>!>!>},<!!!!!>i!!<!>,<u!!{>},{}},{{<!>!>,<!!e>}},{{{{{<!>},<!>,<!<a!!!!!>o}!!{,!>,<!>,<u!>,<u>}},<,!>},<!!}e!}}!!!!!!!>!>!a'!!ae!!}>}},{{}}}},{{{<ai!>},<ie!i>},<,o!>},<!!eie!!!>!>,<<o<a,!>!>a>},{<{!>,<!!!>!!!>a',!>},<>,<!>!!!>o!!!oe!!,!>a!>},<,>}}}}}}";
    println!("final score: {}",parse(input.to_string()));
}